daml-assistant/src/DAML/Assistant/Cache.hs

```diff
@@ -36,63 +37,112 @@ instance Y.FromJSON UpdateCheck where
     -> IO t
     -> IO t
-cacheWith key (CacheTimeout timeout) serialize deserialize damlPath getValue = do
-    let path = cacheFilePath damlPath key
+cacheWith damlPath key timeout ser deser getFresh = do
+    valueAgeM <- loadFromCacheWith damlPath key timeout deser
+    case valueAgeM of
+        Just (value, Fresh) -> pure value
+        Just (value, Stale) -> do
+            valueE <- tryAny getFresh
+            case valueE of
+                Left _ -> pure value
+                Right value' -> do
+                    saveToCacheWith damlPath key ser value'
+                    pure value'
+        Nothing -> do
+            value <- getFresh
+            saveToCacheWith damlPath key ser value
+            pure value
 
+-- | A representation of the age of a cache value. We only care if the value is stale or fresh.
+data CacheAge
+    = Stale
+    | Fresh
+
+-- | Save value to cache.
+saveToCache :: DamlPath -> CacheKey -> String -> IO ()
+saveToCache damlPath key value =
+    void . tryIO $ do
+        let dirPath = cacheDirPath damlPath
+            filePath = cacheFilePath damlPath key
+        createDirectoryIfMissing True dirPath
+        writeFileUTF8 filePath value
+
+-- | Save value to cache, with serialization function. Never raises an exception.
+saveToCacheWith :: DamlPath -> CacheKey -> Serialize t -> t -> IO ()
+saveToCacheWith damlPath key ser value = saveToCache damlPath key (ser value)
+
+-- | Read value from cache, including its age. Never raises an exception.
+loadFromCache :: DamlPath -> CacheKey -> CacheTimeout -> IO (Maybe (String, CacheAge))
+loadFromCache damlPath key (CacheTimeout timeout) = do
+    let path = cacheFilePath damlPath key
     modTimeE <- tryIO (getModificationTime path)
     curTimeE <- tryIO getCurrentTime
-    let useCachedE = liftM2 (\mt ct -> diffUTCTime ct mt < timeout) modTimeE curTimeE
-        useCached = fromRight False useCachedE
-
-    valueMEM <- whenMaybe useCached $ tryIO $ do
-        valueStr <- readFileUTF8 path
-        pure (deserialize valueStr)
-
-    case valueMEM of
-        Just (Right (Just value)) -> pure value
-        _ -> do
-            value <- getValue
-            void . tryIO $ do
-                createDirectoryIfMissing True (cacheDirPath damlPath)
-                writeFileUTF8 path (serialize value)
-            pure value
+    let isStaleE = liftM2 (\mt ct -> diffUTCTime ct mt >= timeout) modTimeE curTimeE
+        isStale  = fromRight True isStaleE
+        age  = if isStale then Stale else Fresh
+    valueM <- eitherToMaybe <$> tryIO (readFileUTF8 path)
+    pure $ fmap (, age) valueM
+
+-- | Read value from cache, including its age, with deserialization function.
+loadFromCacheWith :: DamlPath -> CacheKey -> CacheTimeout -> Deserialize t -> IO (Maybe (t, CacheAge))
+loadFromCacheWith damlPath key timeout deser = do
+    valueAgeM <- loadFromCache damlPath key timeout
+    pure $ do
+        (valueStr, age) <- valueAgeM
+        value <- deser valueStr
```

Comment:
I would just call this `deserialize`. `deser` seems rather confusing.

Reply:
:+1: